source/common/http/codec_client.cc

```diff
@@ -122,7 +122,12 @@ void CodecClient::onReset(ActiveRequest& request, StreamResetReason reason) {
 void CodecClient::onData(Buffer::Instance& data) {
   bool protocol_error = false;
   try {
-    codec_->dispatch(data);
+    auto status = codec_->dispatch(data);
+    if (!status.ok()) {
```

Comment:
Should we differentiate between codec exception and premature response exception here to be clear and maybe assert that it's just the codec type for now if the other is not implemented yet?

Reply:
Hmm. Yes. It looks like I'd need to differentiate between codec/premature, and also have some information on the HttpCode of the premature response. 

We could wrap `absl::Status` to have that data (and have our defined HttpCodes), or we could have an accessor in to the codecs, for eg `getCodecException()`/`getPrematureResponseException`. wdyt? @yanavlasov

Reply:
> We could wrap absl::Status to have that data (and have our defined HttpCodes)

My preference would be for the return value to contain the needed info, similar to the exception it is replacing?

Reply:
Wrapping Status would make it lose some of its value.  Why do we need to distinguish between the two failure conditions?  For protocol error stats?

Reply:
Yeah we do different logic depending on the exception type currently. This needs to be modeled somehow.

Reply:
Looking at this seems like absl::Status might not be what we want. Specifically its error space does not map to Envoy use cases well. Here we use the "Internal error", which is not strictly correct. absl::Status error space is mostly a reflection of internal Google practices centered around stubby and seems to not be well suited for us.
I think we should define our own type very similar to absl::Status but with our own error space, which would let us map exception types naturally. absl::Status is very small if you take various helpers out. We can use the same handle-body pattern to make it fit into a register and get rid of some absl::Status stuff we will not need and perhaps add stuff that we will need.
However, this will require us to make our own StatusOr<> as well.

Reply:
+1 for doing what @yanavlasov suggests if that is the cleanest way of getting what we need. It doesn't seem that bad/hard to duplicate.